onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -1375,6 +1375,105 @@ ORT_API_STATUS_IMPL(OrtApis::GetOpaqueValue, _In_ const char* domain_name, _In_
   return nullptr;
 }
 
+ORT_API_STATUS_IMPL(OrtApis::GetAvailableProviders, _Outptr_ char ***out_ptr,
+                    _In_ int *providers_length) {
+  API_IMPL_BEGIN
+  const int MAX_NUM_PROVIDERS = 14, MAX_LEN = 30;
+  char all_providers[MAX_NUM_PROVIDERS][MAX_LEN] = {
+    "CPUExecutionProvider",
+    "CUDAExecutionProvider",
+    "DnnlExecutionProvider",
+    "NGRAPHExecutionProvider",
+    "OpenVINOExecutionProvider",
+    "NupharExecutionProvider",
+    "VitisAIExecutionProvider",
+    "TensorrtExecutionProvider",
+    "NnapiExecutionProvider",
+    "RknpuExecutionProvider",
+    "DmlExecutionProvider",
+    "MIGraphXExecutionProvider",
+    "ACLExecutionProvider",
+    "ArmNNExecutionProvider",
+  };
+  enum providers {cpu, cuda, dnnl, ngraph, openvino, nuphar, vitisai, tensorrt,
+                  nnapi, rknpu, dml, migraph, acl, armnn};
+  int available_count = 1;
+  enum providers available_providers[MAX_NUM_PROVIDERS];
+  available_providers[available_count-1] = cpu;
+#ifdef USE_CUDA
```

Comment:
It'll be great if we don't have to do these ifdefs in multiple places - python bindings and C API. Can we share this?

Reply:
We could add the checks to a function and call that at both places or just call this function in python bindings. There difference is that python bindings use C++ strings.

Reply:
yes pybind11 is meant to be used with C++ and has built in conversions for C++ strings to python 
not exactly sure how we can share, unless the code in pybind_state.cc has to create C++ strings from the C strings?
also note that the EP string literals used throughout ORT seem to be defined in https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/graph/constants.h

Reply:
I saw constants.h, but didn't see any advantage of using the char * array instead of the string literals here. I have defined the literals with size MAX_LEN, which can be passed to strnlen() and strncpy().

Reply:
We should avoid duplicating the string literals used for the various EPs. They're already defined in constants.h as George said.
I was thinking of a common place that returns the string literal for an EP in the current build. Otherwise, this change will create one more place where we'll have to remember to add the ifdef.

Reply:
Yes, but it doesn't contain an array of EPs. We would still have to remember to add a new char * here if you add a new EP. If we want to have everything in one place, we should define an array of EPs similiar to this in a common place.

Reply:
Yeah and use the string literals from the constants file.

Reply:
I have moved the array of EPs to constants.h.